### PR TITLE
Fixing issue 3206: Label control accessibility reworking

### LINKED ISSUE #3206: Label control does not support UIA accessiblility
* .NET Core Version: 
.NET Core 5.0

 Version:   5.0.100-alpha.1.20073.10
 Commit:    29f4d693a9

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.100-alpha1-05536\

Host (useful for support):
  Version: 5.0.0-alpha.1.20072.3
  Commit:  c3dc1fdfdc

* Have you experienced this same bug with .NET Framework?: 
Yes

**Problem description:**

image

Accessibility providers:
"[pid:31748,providerId:0x11CFA Main:Nested [pid:31900,providerId:0x11CFA Annotation(parent link):Microsoft: Annotation Proxy (unmanaged:uiautomationcore.dll);
Main:Microsoft: MSAA Proxy (unmanaged:uiautomationcore.dll)];
Hwnd(parent link):Microsoft: HWND Proxy (unmanaged:uiautomationcore.dll)]"

**Expected behavior:**
There should be UIA accessibility provider from WinForms

**Minimal repro:**
Create WinForms app with Label control and analyze it using Inspect.

**Related item:**
This work item is the sub of #2164

### PR DESCRIPTION
**Pull request by M-Lipin**
Adding unit tests.

Fixes #3206 

## Proposed changes

- Moving LabelAccessibleObject to a separate file.
- Ensured UIA accessibility support.
- Adding unit tests.

## Customer Impact

- Improving development experience for Label control accessibility. 

## Regression? 

- No

## Risk

- Minimal

### Before

image

### After

image

## Test methodology 

- Manual testing;
- Unit testing;
- UIA automation.

## Test environment(s) 

.NET Core 5.0

 Version:   5.0.100-alpha.1.20073.10
 Commit:    29f4d693a9

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.100-alpha1-05536\

Host (useful for support):
  Version: 5.0.0-alpha.1.20072.3
  Commit:  c3dc1fdfdc

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3208)

### CODE REVIEW

**Comment by M-Lipin** [comment]
This PR is related to #2164 and #3037

File: src/System.Windows.Forms/src/System/Windows/Forms/Label.AccessibleObject.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            }
+
+            private Label OwningLabel => Owner as Label;
+
+            internal override object GetPropertyValue(UiaCore.UIA propertyID)
+                => propertyID switch
+                {
+                    UiaCore.UIA.NamePropertyId => Name,
+                    UiaCore.UIA.AutomationIdPropertyId
+                        => OwningLabel.IsHandleCreated ? OwningLabel.Name : string.Empty,
```
Comment (hughbe) [inline src/System.Windows.Forms/src/System/Windows/Forms/Label.AccessibleObject.cs:42]: Don't we create the `OwningLabel` in the constructor of `Control.ControlAccessibleObject`. Unless `OwningLabel` was disposed, this seems dead. Might be missing something, or you actually want to handle this specific case

File: src/System.Windows.Forms/src/System/Windows/Forms/Label.cs [hunk 2]
```diff
@@ -1643,95 +1643,62 @@ protected override void WndProc(ref Message m)
             }
         }
 
-        [ComVisible(true)]
-        internal class LabelAccessibleObject : ControlAccessibleObject
+        /// <summary>
+        ///  Override ImageList.Indexer to support Label's ImageList semantics.
+        /// </summary>
+        internal class LabelImageIndexer : ImageList.Indexer
```
Comment (hughbe) [inline src/System.Windows.Forms/src/System/Windows/Forms/Label.cs:1649]: nit: could make diff easier to read if you extracted `LabelImageIndexer` to its own class. Not sure though

File: src/System.Windows.Forms/src/System/Windows/Forms/Label.AccessibleObject.cs (hunk 1 above)
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/Label.AccessibleObject.cs:42]: We mustn't create handles for control, if they don't currently have those created. This causes issues like https://github.com/dotnet/winforms/issues/2797

File: src/System.Windows.Forms/src/System/Windows/Forms/Label.cs (hunk 2 above)
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/Label.cs:1649]: Please move this class out, the diff is quite destructive and it is impossible to see what changed.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/LabelAccessibleObjectTests.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+// See the LICENSE file in the project root for more information.
+
+using Xunit;
+
+namespace System.Windows.Forms.Tests
+{
+    public class LabelAccessibleObjectTests
+    {
+        [WinFormsFact]
+        public void LabelAccessibleObject_Constructor_InitializesOwner()
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/LabelAccessibleObjectTests.cs:12]: Need tests to verify handles don't get created

File: src/System.Windows.Forms/src/System/Windows/Forms/Label.cs (hunk 2 above)
Comment (Ryuugamine) [reply src/System.Windows.Forms/src/System/Windows/Forms/Label.cs]: moved to separate file

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/LabelAccessibleObjectTests.cs (hunk 3 above)
Comment (Ryuugamine) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/LabelAccessibleObjectTests.cs:12]: As I understand, handles get created after object construction. I think we can return to this point and investigate it later.

File: src/System.Windows.Forms/src/System/Windows/Forms/Label.AccessibleObject.cs (hunk 1 above)
Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/Label.AccessibleObject.cs:39]: Yes sorry but does `OwningLabel.Name` force handle creation?

File: src/System.Windows.Forms/src/System/Windows/Forms/Label.LabelImageIndexer.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+{
+    public partial class Label
+    {
+        /// <summary>
+        ///  Override ImageList.Indexer to support Label's ImageList semantics.
+        /// </summary>
+        internal class LabelImageIndexer : ImageList.Indexer
+        {
+            private readonly Label? _owner;
+            private bool useIntegerIndex = true;
```
Comment (hughbe) [inline src/System.Windows.Forms/src/System/Windows/Forms/Label.LabelImageIndexer.cs:17]: Not quite related to this PR, but it does seem that `ImageList.Indexer` has a private `_useIntegerIndex` field - some opportunity for future cleanup by expsoing this

File: src/System.Windows.Forms/src/System/Windows/Forms/Label.LabelImageIndexer.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+            public LabelImageIndexer(Label? owner)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Label.LabelImageIndexer.cs:19]: Under which condition can `owner` be `null`?

File: src/System.Windows.Forms/src/System/Windows/Forms/Label.LabelImageIndexer.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+            public override ImageList? ImageList
+            {
+                get { return _owner?.ImageList; }
+                set { Debug.Assert(false, "Setting the image list in this class is not supported"); }
+            }
+
+            public override string? Key
+            {
+                set
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Label.LabelImageIndexer.cs:32]: What happened to the `getter` ❓

File: src/System.Windows.Forms/src/System/Windows/Forms/Label.LabelImageIndexer.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+using System.Diagnostics;
+
+namespace System.Windows.Forms
+{
+    public partial class Label
+    {
+        /// <summary>
+        ///  Override ImageList.Indexer to support Label's ImageList semantics.
+        /// </summary>
+        internal class LabelImageIndexer : ImageList.Indexer
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Label.LabelImageIndexer.cs:14]: Updating or improving this class is not necessary as part of this PR, as it has no relation to the accessibility work.
There's a whole can of worms here, I don't to want to bog down dealing with it.

Please revert this file and keep the class where it was for now.

File: src/System.Windows.Forms/src/System/Windows/Forms/Label.AccessibleObject.cs (hunk 1 above)
Comment (Ryuugamine) [reply src/System.Windows.Forms/src/System/Windows/Forms/Label.AccessibleObject.cs]: Unfortunately, now handle creates after label control was constructed.
But yep, @hughbe right, Name property doesn't force handle creation. I removed IsHandleCreated check and after #3432 it should work fine

**Pull request merged**